rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,327 @@
+
+| Assumption | Behavior | Current option | Repl | Notes |
+|:-----------|:---------|:---------------|:----:|:-----:|
+| `ignoreToPrimitiveHint` | Transform `` `a${x}b` `` to `"a" + x + "b"` instead of `"a".concat(x, "b")` | `loose` in `transform-template-literals` |
+| `mutableTemplateObject` | Don't use `Object.freeze` for the template object created for tagged template literals. This effectively means using the `taggedTemplateLiteralLoose` helper instead of `taggedTemplateLiteral` | `loose` in `transform-template-literals` |
+| `ignoreFunctionLength` | The `.length` of a function should be cropped at the first default argument. When this option is enabled, ignore this spec requirement and don't rely on the `arguments` object | `loose` in `transform-parameters` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABMMAKAhgGkQI0QXkQCZsJsATAxAZgEpEBvAKAF8g&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |
+| `iterableIsArray` | When using an iterable (in array destructuring, for-of or with spreads), assume that the iterable object is an `Array` | `loose` in `transform-destructuring` and `transform-spread`, `assumeArray` in `transform-for-of` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBCELwQ2gOmQIwLoG4BQpLwB4A0EyiAnmrFNkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) (it doesn't support `assumeArray`) | 1 |
+| `arrayLikeIsIterable` | Allow array-like objects to be used where an iterable is expected. This can be useful, for example, to iterate DOM collections in older browsers | `allowArrayLike` in `transform-destructuring`, `transform-spread` and `transform-for-of` | | 1 |
+| `skipForOfIteratorClosing` | When using `for-of` with an iterator, it should always be closed with `.return()` and with `.throw()` in case of an error. This option allows skipping those methods | `loose` in `transform-for-of` | | |
+| `objectRestNoSymbols` | When using rest in object destructuring, don't copy symbol keys. This effectively means using the `objectWithoutPropertiesLoose` helper instead of `objectWithoutProperties` | `loose` in `transform-destructuring` and `proposal-object-rest-spread` _(not documented)_ | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBA3hIDQQHTIEYQL4QLwQMYG4g&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |
```

Comment:
What other could we simplify with this?

Reply:
- `defineEnumerableProperties`
- `objectSpread`
- `objectWithoutProperties`
- Everything that uses `Object.assign` and doesn't check for `getOwnPropertySymbols`
- `instanceof`
- `for await`
- `ignoreToPrimitiveHint`

Reply:
> defineEnumerableProperties
objectSpread
objectWithoutProperties
...

That's really good point! I lean to introduce specific assumptions for each of them. Although a generic assumption ("I don't use symbols") can be handy, it is actually an umbrella term as we may add more optimization assumptions on "I don't use symbols", which could slip into the case of `"loose"` option that we try to avoid in this RFC. Generally the more an assumption is specific, the more easier people can reason about how this assumption might impact the transpiled output.

We can add tags/groups to the assumptions docs instead - So people can filter a list of assumptions under the term "I don't use Symbol".